File: source/Mach1TranscodeEmscripten.js
Module['onRuntimeInitialized'] = function() {
    console.log('loaded Mach1Transcode');

    Module['Mach1Transcode'] = function() {
        this.delete = function() {
            if (this._m1obj) {
                this._m1obj.delete();
                this._m1obj = null;
            }
        };

        this.getInputNumChannels = function() {
            if (this._m1obj) {
                return this._m1obj.getInputNumChannels();
            }
        };

        this.getOutputNumChannels = function() {
            if (this._m1obj) {
                return this._m1obj.getOutputNumChannels();
            }
        };

        this.getFormatFromString = function(str) {
            if (this._m1obj) {
                return this._m1obj.getFormatFromString(str);
            }
        };

        this.getFormatName = function(fmt) {
            if (this._m1obj) {
                return this._m1obj.getFormatName(fmt);
            }
        };

        this.processNormalization = function(bufs) {
            if (this._m1obj) {
                return this._m1obj.processNormalization(bufs);
            }
        };

        this.processMasterGain = function(bufs, masterGain) {
            if (this._m1obj) {
                this._m1obj.processMasterGain(bufs, masterGain);
            }
        };

        this.db2level = function(db) {
            if (this._m1obj) {
                return this._m1obj.db2level(db);
            }
        };

        this.level2db = function(level) {
            if (this._m1obj) {
                return this._m1obj.level2db(level);
            }
        };

        this.setLFESub = function(subChannelIndices, sampleRate) {
            if (this._m1obj) {
                this._m1obj.setLFESub(subChannelIndices, sampleRate);
            }
        };

        this.setSpatialDownmixer = function(corrThreshold) {
            if (this._m1obj) {
                this._m1obj.setSpatialDownmixer(corrThreshold);
            }
        };

        this.getSpatialDownmixerPossibility = function() {
            if (this._m1obj) {
                return this._m1obj.getSpatialDownmixerPossibility();
            }
        };

        this.setInputFormat = function(inFmt) {
            if (this._m1obj) {
                this._m1obj.setInputFormat(inFmt);
            }
        };

        this.setInputFormatCustomPointsJson = function(inJson) {
            if (this._m1obj) {
                this._m1obj.setInputFormatTTJson(inJson);
            }
        };

        this.setInputFormatCustomPoints = function(points) {
            if (this._m1obj) {
                this._m1obj.setInputFormatTTPoints(points);
            }
        };

        this.setOutputFormat = function(outFmt) {
            if (this._m1obj) {
                this._m1obj.setOutputFormat(outFmt);
            }
        };

        this.setOutputFormatCustomPointsJson = function(outJson) {
            if (this._m1obj) {
                this._m1obj.setOutputFormatTTJson(outJson);
            }
        };

        this.setOutputFormatCustomPoints = function(points) {
            if (this._m1obj) {
                this._m1obj.setOutputFormatTTPoints(points);
            }
        };

        this.processConversionPath = function() {
            if (this._m1obj) {
                return this._m1obj.processConversionPath();
            }
        };

        this.getMatrixConversion = function() {
            if (this._m1obj) {
                var obj = this._m1obj.getMatrixConversion();
                var arr = [];
                for (var i = 0; i < obj.size(); i++) {
                    arr[i] = [];
                    var obj2 = obj.get(i);
                    for (var j = 0; j < obj2.size(); j++) {
                        arr[i][j] = obj2.get(j);
                    }

                }
                return arr;
            }
        };

        this.processConversion = function(inBufs, outBufs) {
            if (this._m1obj) {
                return this._m1obj.processConversion(inBufs, outBufs);
            }
        };

        this.getFormatConversionPath = function() {
            if (this._m1obj) {
                var obj = this._m1obj.getFormatConversionPath();
                var arr = [];
                for (var i = 0; i < obj.size(); i++) {
                    arr[i] = obj.get(i);
                }
                return arr;
            }
        };

        this._m1obj = new(Module).Mach1TranscodeInternal();
    };
};

File: libmach1spatial/api_transcode/src/Mach1TranscodeEmscripten.cpp
#include <emscripten/bind.h>

#include "../include/Mach1Transcode.h"
#include "Mach1TranscodeCAPI.h"

using namespace emscripten;

EMSCRIPTEN_BINDINGS(Mach1Transcode) {
    register_vector<float>("VectorFloat");
    register_vector<std::string>("VectorString");
    register_vector<Mach1Point3D>("VectorMach1Point3D");
    register_vector<std::vector<float> >("VectorVectorFloat");

    value_object<Mach1Point3D>("Mach1Point3D")
        .field("x", &Mach1Point3D::x)
        .field("y", &Mach1Point3D::y)
        .field("z", &Mach1Point3D::z);

    class_<Mach1Transcode<float> >("Mach1TranscodeInternal")
        .constructor<>()
        .function("getInputNumChannels", &Mach1Transcode<float>::getInputNumChannels)
        .function("getOutputNumChannels", &Mach1Transcode<float>::getOutputNumChannels)
        .function("getFormatFromString", &Mach1Transcode<float>::getFormatFromString)
        .function("getFormatName", &Mach1Transcode<float>::getFormatName)
        .function("getAllFormatNames", &Mach1Transcode<float>::getAllFormatNames)

        .function("processNormalization", &Mach1Transcode<float>::processNormalization)
        .function("processMasterGain", &Mach1Transcode<float>::processMasterGain)

        .function("db2level", &Mach1Transcode<float>::db2level)
        .function("level2db", &Mach1Transcode<float>::level2db)

        .function("setLFESub", &Mach1Transcode<float>::setLFESub)
        .function("setSpatialDownmixer", &Mach1Transcode<float>::setSpatialDownmixer)
        .function("getSpatialDownmixerPossibility", &Mach1Transcode<float>::getSpatialDownmixerPossibility)

        .function("setInputFormat", &Mach1Transcode<float>::setInputFormat)
        .function("setInputFormatCustomPointsJson", &Mach1Transcode<float>::setInputFormatCustomPointsJson)
        .function("setInputFormatCustomPoints", &Mach1Transcode<float>::setInputFormatCustomPoints)

        .function("setOutputFormat", &Mach1Transcode<float>::setOutputFormat)
        .function("setOutputFormatCustomPointsJson", &Mach1Transcode<float>::setOutputFormatCustomPointsJson)
        .function("setOutputFormatCustomPoints", &Mach1Transcode<float>::setOutputFormatCustomPoints)

        .function("processConversionPath", &Mach1Transcode<float>::processConversionPath)
        .function("getMatrixConversion", &Mach1Transcode<float>::getMatrixConversion)
        .function("processConversion", &Mach1Transcode<float>::processConversion)

        .function("getFormatConversionPath", &Mach1Transcode<float>::getFormatConversionPath);
}
